tower-balance/src/load/peak_ewma.rs

```diff
@@ -0,0 +1,212 @@
+            node: Arc::new(Mutex::new(Node::new(decay))),
+            _p: PhantomData,
+        }
+    }
+
+    fn instrument(&self) -> Instrument {
+        Instrument {
+            start: Instant::now(),
+            node: self.node.clone(),
+        }
+    }
+}
+
+impl<S, M> Service for PeakEWMA<S, M>
+where
+    S: Service,
+    M: Measure<Instrument, S::Response>,
+{
+    type Request = S::Request;
+    type Response = M::Measured;
+    type Error = S::Error;
+    type Future = MeasureFuture<S::Future, M, Instrument>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.service.poll_ready()
+    }
+
+    fn call(&mut self, req: Self::Request) -> Self::Future {
+        MeasureFuture::new(self.instrument(), self.service.call(req))
+    }
+}
+
+const PENALTY: f64 = 65535.0;
+
+impl<S, M> Load for PeakEWMA<S, M> {
+    type Metric = Cost;
+
+    fn load(&self) -> Self::Metric {
+        let pending = Arc::strong_count(&self.node) as u32 - 1;
+        let mut node = self.node.lock().expect("peak ewma state");
+        if node.rtt_estimate == 0.0 && pending > 0 {
+            debug!("penalty={} pending={}", PENALTY, pending);
+            Cost(PENALTY + f64::from(pending))
+        } else {
+            // Update the cost to account for decay since the last update.
+            let rtt_estimate = node.update_rtt_estimate(0.0);
+
+            let cost = rtt_estimate * (f64::from(pending) + 1.0);
+            debug!(
+                "load estimate={:.0}ms pending={} cost={:.0}",
+                node.rtt_estimate / 1_000_000.0,
+                pending,
+                cost
+            );
+            Cost(cost)
+        }
+    }
+}
+
+// ===== impl Node =====
+
+impl Node {
+    fn new(decay: Duration) -> Self {
+        Self {
+            tau: nanos(decay),
+            last_update: Instant::now(),
+            rtt_estimate: 0.0,
+        }
+    }
+
+    fn update_rtt_estimate(&mut self, rtt: f64) -> f64 {
+        let now = Instant::now();
+        if self.rtt_estimate < rtt {
+            trace!("update peak rtt={}", rtt);
+            self.rtt_estimate = rtt;
+        } else {
+            // Determine how heavily to weight our rtt estimate.
+            let td = nanos(now - self.last_update);
+            let recency = (-td / self.tau).exp();
+
+            let e = (self.rtt_estimate * recency) + (rtt * (1.0 - recency));
+            trace!(
+                "update prior={:.0}ms * {:.5} + rtt={:.0}ms * {:.5}; estimate={:.0}ms",
+                self.rtt_estimate / 1_000_000.0,
+                recency,
+                rtt / 1_000_000.0,
+                1.0 - recency,
+                e / 1_000_000.0
+            );
+            self.rtt_estimate = e;
+        }
+        self.last_update = now;
+        self.rtt_estimate
+    }
+}
+
+// ===== impl Instrument =====
+
+impl Drop for Instrument {
+    fn drop(&mut self) {
+        if let Ok(mut node) = self.node.lock() {
+            let rtt = self.start.elapsed();
```

Comment:
Is it intended for the elapsed computation to occur within the lock's critical section?